casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -483,74 +484,95 @@ object Validate {
+        if (validator == b.sender) {
+          // We let checkEquivocations handle this case
+          Applicative[F].pure(Left(tail))
+        } else {
+          val previousBlockJustificationHash =
+            previousLatestMessages.getOrElse(
+              validator,
+              genesis.blockHash
+            )
+          Monad[F].ifM(
+            isJustificationRegression[F](
+              dag,
+              currentBlockJustificationHash,
+              previousBlockJustificationHash
+            )
+          )(
+            {
+              val message =
+                s"block ${PrettyPrinter.buildString(currentBlockJustificationHash)} by ${PrettyPrinter
+                  .buildString(validator)} has a lower sequence number than ${PrettyPrinter.buildString(previousBlockJustificationHash)}."
+              Log[F].warn(ignore(b, message)) *> Applicative[F].pure(
+                Right(Left(JustificationRegression))
+              )
+            },
+            Applicative[F].pure(Left(tail))
+          )
+        }
+    }
 
-  private def isJustificationRegression[F[_]: Monad: Log: BlockStore](
+  private def isJustificationRegression[F[_]: Sync](
+      dag: BlockDagRepresentation[F],
       currentBlockJustificationHash: BlockHash,
       previousBlockJustificationHash: BlockHash
   ): F[Boolean] =
     for {
-      currentBlockJustification <- ProtoUtil
-                                    .unsafeGetBlock[F](currentBlockJustificationHash)
-      previousBlockJustification <- ProtoUtil
-                                     .unsafeGetBlock[F](previousBlockJustificationHash)
+      maybeCurrentBlockJustification <- dag.lookup(currentBlockJustificationHash)
+      currentBlockJustification <- maybeCurrentBlockJustification match {
+                                    case Some(block) => block.pure[F]
+                                    case None =>
+                                      Sync[F].raiseError[BlockMetadata](
+                                        new Exception(
+                                          s"Missing ${PrettyPrinter.buildString(currentBlockJustificationHash)} from block dag store."
+                                        )
+                                      )
+                                  }
+      maybePreviousBlockJustification <- dag.lookup(previousBlockJustificationHash)
+      previousBlockJustification <- maybePreviousBlockJustification match {
+                                     case Some(block) => block.pure[F]
+                                     case None =>
+                                       Sync[F].raiseError[BlockMetadata](
+                                         new Exception(
+                                           s"Missing ${PrettyPrinter.buildString(previousBlockJustificationHash)} from block dag store."
+                                         )
+                                       )
+                                   }
```

Comment:
maybe deduplicate?

Reply:
I agree. You have used this pattern a few times in the previous PRs. Maybe it would make sense to create a utility method just like `ProtoUtil.unsafeGetBlock` but using block DAG storage instead of block storage (or even change the `ProtoUtil.unsafeGetBlock` itself).

Reply:
Yeah I was thinking I would eventually push the unsafeGetBlock to the block dag store level.